File: leet-code/educative.io/07-tree-breadth-first-search/average-of-levels-in-binary-tree.cpp
// https://leetcode.com/problems/average-of-levels-in-binary-tree/

using namespace std;

#include <queue>
#include <vector>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
};

class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> result;
        if (root == nullptr) {
            return result;
        }
        
        // add the root into queue
        queue<TreeNode *>queue;
        queue.push(root);
        
        while (!queue.empty()) {
            int levelSize = queue.size();
            double sum = 0.0;
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode *currentNode = queue.front();
                queue.pop();
                
                // sum the node value
                sum += currentNode->val;
                
                // insert the children of current node in the queue
                if (currentNode->left != nullptr) {
                    queue.push(currentNode->left);
                }
                if (currentNode->right != nullptr) {
                    queue.push(currentNode->right);
                }
            }
            result.push_back(sum / levelSize);
        }
        
        return result;
    }
};

File: leet-code/leetcoding-challenge/2021-mar/week1/average-of-levels-in-binary-tree.py
# https://leetcode.com/problems/average-of-levels-in-binary-tree/

from collections import deque
from typing import List

"""
Topic: Breadth First Search
"""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val: int = 0, left: int = None, right: int = None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        queue = deque()
        queue.append(root)
        result = []

        while queue:
            levelSize = len(queue)
            total = 0.0
            
            for i in range(levelSize):
                node = queue.popleft()
                total += node.val
                
                if node.left != None:
                    queue.append(node.left)
                if node.right != None:
                    queue.append(node.right)
            
            result.append(total / levelSize)

        return result
